src/VisualStudio/Xaml/Impl/Implementation/XamlTextViewCreationListener.cs

```diff
@@ -0,0 +1,185 @@
+            this.vsSolution = (IVsSolution)serviceProvider.GetService(typeof(SVsSolution));
+
+            uint solutionEventsCookie;
+            if (ErrorHandler.Succeeded(this.vsSolution.AdviseSolutionEvents(this, out solutionEventsCookie)))
+            {
+                this.solutionEventsCookie = solutionEventsCookie;
+            }
+        }
+
+        public void VsTextViewCreated(IVsTextView vsTextView)
+        {
+            if (vsTextView == null)
+            {
+                throw new ArgumentNullException(nameof(vsTextView));
+            }
+
+            IVsTextLines textLines;
+            if (ErrorHandler.Failed(vsTextView.GetBuffer(out textLines)))
+            {
+                return;
+            }
+
+            IVsUserData userData = textLines as IVsUserData;
+            if (userData == null)
+            {
+                return;
+            }
+
+            object monikerObj;
+            Guid monikerGuid = typeof(IVsUserData).GUID;
+            if (ErrorHandler.Failed(userData.GetData(ref monikerGuid, out monikerObj)))
+            {
+                return;
+            }
+
+            string filePath = monikerObj as string;
+            uint itemId;
+            uint docCookie;
+            IVsHierarchy hierarchy;
+
+            this.rdt.Value.FindDocument(filePath, out hierarchy, out itemId, out docCookie);
+
+            AbstractProject project = GetXamlProject(hierarchy);
+            if (project == null)
+            {
+                project = new XamlProject(hierarchy, this.serviceProvider, this.vsWorkspace);
+                this.vsWorkspace.ProjectTracker.AddProject(project);
+            }
+
+            IVisualStudioHostDocument vsDocument = this.GetXamlDocument(project, filePath);
+            if (vsDocument == null)
+            {
+                return;
+            }
+
+            this.AttachRunningDocTableEvents();
+
+            project.AddDocument(vsDocument, isCurrentContext: true);
+
+            IWpfTextView wpfTextView = this.editorAdaptersFactory.GetWpfTextView(vsTextView);
+            XamlOleCommandTarget target = new XamlOleCommandTarget(wpfTextView, this.CommandHandlerServiceFactory, this.editorAdaptersFactory, this.serviceProvider);
+            target.AttachToVsTextView();
+        }
+
+        private void AttachRunningDocTableEvents()
+        {
+            if (!this.rdtEventsCookie.HasValue)
+            {
+                this.rdtEventsCookie = this.rdt.Value.Advise(this);
+            }
+        }
+
+        private AbstractProject GetXamlProject(IVsHierarchy hierarchy)
+        {
+            return this.vsWorkspace.ProjectTracker.Projects.FirstOrDefault(p => p.Language == StringConstants.XamlLanguageName && p.Hierarchy == hierarchy);
+        }
+
+        private IVisualStudioHostDocument GetXamlDocument(AbstractProject project, string filePath, bool create = true)
```

Comment:
We would tend *not* to have an optional param for a non-public API unless one use was *vastly* more common, otherwise, we fall on the side of making the caller think about it and make a choice.